File: ios/Truemap/Services/ReconstructionService.swift
import Foundation
import Combine

/// Processing status
enum ProcessingStatus: Int {
    case idle = 0
    case downloadingVideo
    case extractingFrames
    case extractingFeatures
    case matchingFeatures
    case computingSfM
    case aligningModels
    case bundleAdjustment
    case denseReconstruction
    case generatingMesh
    case exporting
    case completed
    case error
}

/// Progress update
struct ProgressUpdate {
    let status: ProcessingStatus
    let progress: Double
    let message: String
}

/// Reconstruction service that interfaces with C++ core
class ReconstructionService {
    private let progressSubject = PassthroughSubject<ProgressUpdate, Never>()
    private let statusSubject = CurrentValueSubject<ProcessingStatus, Never>(.idle)
    
    var progressPublisher: AnyPublisher<ProgressUpdate, Never> {
        progressSubject.eraseToAnyPublisher()
    }
    
    var statusPublisher: AnyPublisher<ProcessingStatus, Never> {
        statusSubject.eraseToAnyPublisher()
    }
    
    // Bridge to C++ (would be implemented via Objective-C++)
    private var coreHandle: OpaquePointer?
    
    init() {
        // Initialize C++ core
        // coreHandle = truemap_core_create()
    }
    
    deinit {
        // Clean up C++ core
        // if let handle = coreHandle {
        //     truemap_core_destroy(handle)
        // }
    }
    
    func startReconstruction(urls: [String], completion: @escaping (Result<Void, Error>) -> Void) {
        // This would call into C++ implementation
        // For now, simulate the process
        
        statusSubject.send(.downloadingVideo)
        
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }
            
            // Simulate processing stages
            let stages: [(ProcessingStatus, String)] = [
                (.downloadingVideo, "Downloading videos..."),
                (.extractingFrames, "Extracting frames..."),
                (.extractingFeatures, "Extracting features..."),
                (.matchingFeatures, "Matching features..."),
                (.computingSfM, "Computing Structure from Motion..."),
                (.aligningModels, "Aligning sub-models..."),
                (.bundleAdjustment, "Running bundle adjustment..."),
                (.denseReconstruction, "Generating dense point cloud..."),
                (.generatingMesh, "Creating mesh..."),
            ]
            
            for (index, stage) in stages.enumerated() {
                Thread.sleep(forTimeInterval: 0.5) // Simulate work
                
                let progress = Double(index + 1) / Double(stages.count)
                let update = ProgressUpdate(
                    status: stage.0,
                    progress: progress,
                    message: stage.1
                )
                
                self.progressSubject.send(update)
                self.statusSubject.send(stage.0)
            }
            
            self.statusSubject.send(.completed)
            self.progressSubject.send(ProgressUpdate(
                status: .completed,
                progress: 1.0,
                message: "Reconstruction completed!"
            ))
            
            completion(.success(()))
        }
    }
    
    func cancelReconstruction() {
        // Call C++ cancel function
        // truemap_core_cancel(coreHandle)
        statusSubject.send(.idle)
    }
    
    func exportMesh(completion: @escaping (Result<URL, Error>) -> Void) {
        // Export mesh to file
        let documentsPath = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let outputURL = documentsPath.appendingPathComponent("truemap_mesh.obj")
        
        // Call C++ export function
        // truemap_core_export_mesh(coreHandle, outputURL.path)
        
        DispatchQueue.global(qos: .userInitiated).asyncAfter(deadline: .now() + 0.5) {
            completion(.success(outputURL))
        }
    }
}

File: android/app/src/main/kotlin/com/truemap/app/ReconstructionService.kt
package com.truemap.app

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext

enum class ProcessingStatus {
    IDLE,
    DOWNLOADING_VIDEO,
    EXTRACTING_FRAMES,
    EXTRACTING_FEATURES,
    MATCHING_FEATURES,
    COMPUTING_SFM,
    ALIGNING_MODELS,
    BUNDLE_ADJUSTMENT,
    DENSE_RECONSTRUCTION,
    GENERATING_MESH,
    EXPORTING,
    COMPLETED,
    ERROR
}

class ReconstructionService {
    private var progressCallback: ((ProcessingStatus, Double, String) -> Unit)? = null
    private var shouldCancel = false
    
    // Native methods (JNI bridge to C++)
    private external fun nativeInit(): Long
    private external fun nativeDestroy(handle: Long)
    private external fun nativeAddVideoUrl(handle: Long, url: String): Boolean
    private external fun nativeStartReconstruction(handle: Long): Boolean
    private external fun nativeCancel(handle: Long)
    private external fun nativeExportMesh(handle: Long, path: String): Boolean
    
    companion object {
        init {
            // Load native library
            // System.loadLibrary("truemap_core")
        }
    }
    
    fun setProgressCallback(callback: (ProcessingStatus, Double, String) -> Unit) {
        progressCallback = callback
    }
    
    suspend fun startReconstruction(urls: List<String>) = withContext(Dispatchers.IO) {
        shouldCancel = false
        
        // In real implementation, this would call native C++ code
        // For now, simulate the process
        
        val stages = listOf(
            ProcessingStatus.DOWNLOADING_VIDEO to "Downloading videos...",
            ProcessingStatus.EXTRACTING_FRAMES to "Extracting frames...",
            ProcessingStatus.EXTRACTING_FEATURES to "Extracting features...",
            ProcessingStatus.MATCHING_FEATURES to "Matching features...",
            ProcessingStatus.COMPUTING_SFM to "Computing Structure from Motion...",
            ProcessingStatus.ALIGNING_MODELS to "Aligning sub-models...",
            ProcessingStatus.BUNDLE_ADJUSTMENT to "Running bundle adjustment...",
            ProcessingStatus.DENSE_RECONSTRUCTION to "Generating dense point cloud...",
            ProcessingStatus.GENERATING_MESH to "Creating mesh..."
        )
        
        for ((index, stage) in stages.withIndex()) {
            if (shouldCancel) {
                progressCallback?.invoke(ProcessingStatus.IDLE, 0.0, "Cancelled")
                return@withContext
            }
            
            delay(500) // Simulate work
            
            val progress = (index + 1).toDouble() / stages.size
            progressCallback?.invoke(stage.first, progress, stage.second)
        }
        
        progressCallback?.invoke(ProcessingStatus.COMPLETED, 1.0, "Reconstruction completed!")
    }
    
    fun cancelReconstruction() {
        shouldCancel = true
        // nativeCancel(handle)
    }
    
    suspend fun exportMesh(): String = withContext(Dispatchers.IO) {
        // Export to app's files directory
        val outputPath = "/storage/emulated/0/Download/truemap_mesh.obj"
        
        // In real implementation:
        // nativeExportMesh(handle, outputPath)
        
        delay(500) // Simulate export
        
        return@withContext outputPath
    }
}
